### ISSUE TITLE
Nested anonymous subs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6693]
Migrated from [rt.perl.org#23265](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=23265) (status was 'resolved')

Searchable as RT23265$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by .lmai@web.de


Good morning\!

\#\!perl
use warnings;
use strict;
my $tests = 2;
print "1\.\.$tests\\n"\,
&nbsp;    sub \{$\_\[0\]\->\(@&#8203;\_\)\} \-> \(
&nbsp;        sub \{
&nbsp;            $\_\[1\] ?
&nbsp;            $\_\[0\]\->\($\_\[0\]\, $\_\[1\] \- 1\) \.
&nbsp;            sub \{"ok $\_\[0\]\\n"\}\->\($\_\[1\]\) :
&nbsp;            ''
&nbsp;        \}\,
&nbsp;        $tests
&nbsp;    \)
;
\_\_END\_\_

This script generates the expected output&#8203;:
1\.\.2
ok 1
ok 2

But perl\-5\.8\.1\-RC4 continues with&#8203;:
Attempt to free unreferenced scalar\.
Segmentation fault \(core dumped\)

My debugger tells me&#8203;:
\#0  0x080df580 in Perl\_sv\_free \(sv=0x1\) at sv\.c&#8203;:5305
5305        if \(SvREFCNT\(sv\) == 0\) \{
\(gdb\) bt
\#0  0x080df580 in Perl\_sv\_free \(sv=0x1\) at sv\.c&#8203;:5305
\#1  0x0809cd15 in Perl\_pad\_undef \(cv=0x817d410\) at pad\.c&#8203;:278
\#2  0x08094f45 in Perl\_cv\_undef \(cv=0x817d410\) at op\.c&#8203;:3942
\#3  0x080df22e in Perl\_sv\_clear \(sv=0x817d410\) at sv\.c&#8203;:5154
\#4  0x080df71d in Perl\_sv\_free \(sv=0x817d410\) at sv\.c&#8203;:5338
\#5  0x0805fdbc in perl\_destruct \(my\_perl=0x817d148\) at perl\.c&#8203;:473
\#6  0x0805e8dd in main \(argc=2\, argv=0xbffff7f4\, env=0xbffff800\)
&nbsp;    at perlmain\.c&#8203;:88
\#7  0x400a9baf in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6

HTH
Lukas Mai

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.1:

Configured by lmai at Fri Aug  8 02:39:51 CEST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.2.18, archname=i686-linux-64int-ld
    uname='linux barubary 2.2.18 #1 fri jan 19 22:10:35 gmt 2001 i686 unknown '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -
DDEBUGGING',
    optimize='-g',
    cppflags='-fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING'
    ccversion='', gccversion='3.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldbm -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    RC4


@INC for perl v5.8.1:
    perl-5.8.1-RC4/lib
    /usr/local/lib/perl5/5.8.1/i686-linux-64int-ld
    /usr/local/lib/perl5/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.1/i686-linux-64int-ld
    /usr/local/lib/perl5/site_perl/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.0/i686-linux-64int-ld
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.1:
    HOME=/home/lmai
    LANG=en_US
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=/home/lmai/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/lib/java/bin:/var/lib/dosemu:/usr/games
/bin:/usr/games:/opt/gnome/bin:/opt/kde2/bin:/opt/kde/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Sat\, Aug 09\, 2003 a\.d\.\, Lukas Mai wrote&#8203;:
> \#\!perl
> use warnings;
> use strict;
> my $tests = 2;
> print "1\.\.$tests\\n"\,
>     sub \{$\_\[0\]\->\(@&#8203;\_\)\} \-> \(
>         sub \{
>             $\_\[1\] ?
>             $\_\[0\]\->\($\_\[0\]\, $\_\[1\] \- 1\) \.
>             sub \{"ok $\_\[0\]\\n"\}\->\($\_\[1\]\) :
>             ''
>         \}\,
>         $tests
>     \)
> ;
> \_\_END\_\_
> 
> This script generates the expected output&#8203;:
> 1\.\.2
> ok 1
> ok 2
> 
> But perl\-5\.8\.1\-RC4 continues with&#8203;:
> Attempt to free unreferenced scalar\.
> Segmentation fault \(core dumped\)

This rather dodges than fixes&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- /arc/bleadperl/pad.c	2003-08-09 11:40:37.000000000 +0300
+++ ./pad.c	2003-08-09 22:53:58.000000000 +0300
@@ -254,7 +254,8 @@
 		curpad[ix] = Nullsv;
 		SvREFCNT_dec(innercv);
 		if (SvREFCNT(innercv) /* in use, not just a prototype */
-		    && CvOUTSIDE(innercv) == cv)
+		    && CvOUTSIDE(innercv) == cv
+		    && SvREFCNT(outercv))
 		{
 		    assert(CvWEAKOUTSIDE(innercv));
 		    CvWEAKOUTSIDE_off(innercv);


```
</details>
Just thinking about Perl's global destruction make me shudder :-)

Regards\,
Adi

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Aug 09\, 2003 at 11&#8203;:00&#8203;:39PM \+0300\, Enache Adrian wrote&#8203;:
> On Sat\, Aug 09\, 2003 a\.d\.\, Lukas Mai wrote&#8203;:
> > \#\!perl
> > use warnings;
> > use strict;
> > my $tests = 2;
> > print "1\.\.$tests\\n"\,
> >     sub \{$\_\[0\]\->\(@&#8203;\_\)\} \-> \(
> >         sub \{
> >             $\_\[1\] ?
> >             $\_\[0\]\->\($\_\[0\]\, $\_\[1\] \- 1\) \.
> >             sub \{"ok $\_\[0\]\\n"\}\->\($\_\[1\]\) :
> >             ''
> >         \}\,
> >         $tests
> >     \)
> > ;
> > \_\_END\_\_
> > 
> > This script generates the expected output&#8203;:
> > 1\.\.2
> > ok 1
> > ok 2
> > 
> > But perl\-5\.8\.1\-RC4 continues with&#8203;:
> > Attempt to free unreferenced scalar\.
> > Segmentation fault \(core dumped\)
> 
> This rather dodges than fixes&#8203;:

and this fixes rather than dodges :\-\)

Dave\.

\-\- 
"Emacs isn't a bad OS once you get used to it\.
It just lacks a decent editor\."

<details><summary>Inline Patch</summary>

```diff
--- pad.c-	Sun Aug 10 01:13:29 2003
+++ pad.c	Sun Aug 10 01:15:57 2003
@@ -258,9 +258,15 @@ Perl_pad_undef(pTHX_ CV* cv)
 		{
 		    assert(CvWEAKOUTSIDE(innercv));
 		    CvWEAKOUTSIDE_off(innercv);
-		    CvOUTSIDE(innercv) = outercv;
 		    CvOUTSIDE_SEQ(innercv) = seq;
-		    SvREFCNT_inc(outercv);
+		    /* don't relink to grandfather if he's being freed */
+		    if (SvREFCNT(outercv)) {
+			CvOUTSIDE(innercv) = outercv;
+			SvREFCNT_inc(outercv);
+		    }
+		    else {
+			CvOUTSIDE(innercv) = Nullcv;
+		    }
 		}
 	    }
 	}
--- t/op/closure.t-	Sun Aug 10 01:17:11 2003
+++ t/op/closure.t	Sun Aug 10 01:31:38 2003
@@ -13,7 +13,7 @@
 
 use Config;
 
-print "1..184\n";
+print "1..185\n";
 
 my $test = 1;
 sub test (&) {
@@ -641,4 +641,27 @@
     test { $a{7}->()->() + $a{11}->()->() == 18 };
 }
 
+# bugid #23265 - this used to coredump during destruction of PL_maincv
+# and its children
+
+require './test.pl';
+
+my $got = runperl(
+    prog => q[
+	print
+	    sub {$_[0]->(@_)} -> (
+		sub {
+		    $_[1]
+			?  $_[0]->($_[0], $_[1] - 1) .  sub {"x"}->()
+			: "y"
+		},   
+		2
+	    )
+	    , "\n"
+	;            
+    
+    ],
+    stderr => 1
+);
+test { $got eq "yxx\n" };
 
```
</details>

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Sun\, Aug 10\, 2003 a\.d\.\, Dave Mitchell wrote&#8203;:
>              assert\(CvWEAKOUTSIDE\(innercv\)\);
>              CvWEAKOUTSIDE\_off\(innercv\);
> \-            CvOUTSIDE\(innercv\) = outercv;
>              CvOUTSIDE\_SEQ\(innercv\) = seq;
> \-            SvREFCNT\_inc\(outercv\);
> \+            /\* don't relink to grandfather if he's being freed \*/
> \+            if \(SvREFCNT\(outercv\)\) \{
> \+            CvOUTSIDE\(innercv\) = outercv;
> \+            SvREFCNT\_inc\(outercv\);
> \+            \}
> \+            else \{
> \+            CvOUTSIDE\(innercv\) = Nullcv;
> \+            \}

I'm not sure I get the difference :\-\)
Why should be CvOUTSIDE\_SEQ\, CvWEAKOUTSIDE set/cleared if the outside
CV is going to be NULL ?

Regards\,
Adi

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Aug 10\, 2003 at 04&#8203;:16&#8203;:49AM \+0300\, Enache Adrian wrote&#8203;:
> On Sun\, Aug 10\, 2003 a\.d\.\, Dave Mitchell wrote&#8203;:
> >              assert\(CvWEAKOUTSIDE\(innercv\)\);
> >              CvWEAKOUTSIDE\_off\(innercv\);
> > \-            CvOUTSIDE\(innercv\) = outercv;
> >              CvOUTSIDE\_SEQ\(innercv\) = seq;
> > \-            SvREFCNT\_inc\(outercv\);
> > \+            /\* don't relink to grandfather if he's being freed \*/
> > \+            if \(SvREFCNT\(outercv\)\) \{
> > \+            CvOUTSIDE\(innercv\) = outercv;
> > \+            SvREFCNT\_inc\(outercv\);
> > \+            \}
> > \+            else \{
> > \+            CvOUTSIDE\(innercv\) = Nullcv;
> > \+            \}
> 
> I'm not sure I get the difference :\-\)

Mainly that my patch sets CvOUTSIDE\(\) to null rather than leaving it
dangling\.

> Why should be CvOUTSIDE\_SEQ\, CvWEAKOUTSIDE set/cleared if the outside
> CV is going to be NULL ?

No particular reason\. I was being lazy\.

\-\- 
In my day\, we used to edit the inodes by hand\.  With magnets\.

**Comment by p5pRT** [comment]
### From @rgs

Dave's patch applied to 5\.9\.0 as change 20597\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

> Dave's patch applied to 5\.9\.0 as change 20597\.

But retracted later\. Reopen\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

and fixed again by patch \#20631

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
